### PR TITLE
18% performance boost in Nef_3 K3_tree.

### PR DESCRIPTION
**Pull request by GilesBathgate**
## Summary of Changes

Looking carefully at the code in K3_tree and running it through a profiler it can be seen that the code firstly creates an `Object_list` of vertices, edges, and facets. It then spends substantial amount of time unpicking the object list, and casting each object back into vertices/edges/facets to build the K3_tree. While the casts themselves are not very expensive, applying the `make_object` and `CGAL::assign` lavishly so many items does add up to significant overhead. By simply passing separate lists of Vertices, Edges, and Facets much of this casting is avoided and the performance can be boosted. Using the Nef_3_benchmark with spheregrid/shiftedspheregrid a performance gain of 10% can be observed. The Object lists are created at the point a K3_tree `Node` is created, so as to be backwards compatible with the existing code.

Another improvement is to select a splitting plane at the midpoint of the median and its predecessor, this ensures fewer points fall on the oriented boundary and gives a further performance boost of 8%

A few other minor tweaks and cleaning has been applied to the code, for which further details can be discussed if requested.

## Release Management

* Affected package(s): Nef_3
* Issue(s) solved (if any): performance
* License and copyright ownership: Returned to CGAL authors.

### CODE REVIEW

File: Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h
Comment (afabri) [inline Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:111]: Hi Giles, I guess you are not really happy that we are lagging behind with approving and merging your tsunami of pull requests.  We really appreciate them and we allocate resources where we find time. 
The problem with this PR is that it seems to combine several others in the diff. That makes it pretty hard to review.  I suggest that we deal with the PRs in order and when an older one  (e.g. the one on "triangulate faces") gets merged into mater this block should be no longer in the diff.

File: Nef_3/include/CGAL/Nef_3/K3_tree.h
Comment (afabri) [inline Nef_3/include/CGAL/Nef_3/K3_tree.h:384]: I am puzzled to see `Depth`  as template parameter.  Wouldn't that need a constant known at compile time. @lrineau @mglisse

Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:384]: Its always an int, its specialised by a parameter `build_k3tree(O, O.size()+v_end, 0)`

File: Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h
Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:111]: Yeah, I was finding it tricky to manage. If you are happy to hold off fully reviewing this until #5740 is merged, I can rebase and resolve conflicts.

File: Nef_3/include/CGAL/Nef_3/K3_tree.h
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:384]: But isn't `O.size()+v_end`  only known at runtime?

Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:384]: No, it's the final parameter, i.e the `0` which is a literal int.

File: Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h
Comment (GilesBathgate) [inline Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:60]: @afabri I think this template specialisation (and also for Smaller_than) should be removed because:

a) It doesn't seem to match against Epeck kernels.
b) I think CGAL::compare_x already does an approx comparison when possible.

File: Nef_3/include/CGAL/Nef_3/K3_tree.h
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:384]: I misunderstood. I thought we have to do with for example `Depth-1`  being used in a recursion, but here it is just an integral type in the parameter list. So I agree with this change.

File: Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:60]: I agree. I am even wondering if we should not add the 3D equivalent to  [`Compare_x_2`](https://github.com/CGAL/cgal/blob/master/Filtered_kernel/include/CGAL/internal/Static_filters/Compare_x_2.h).

**Comment by afabri** [comment]
Is this still WIP or shall I review it?

**Comment by GilesBathgate** [comment]
> Is this still WIP or shall I review it?

The conflicts regarding CGAL_NEF3_TRIANGULATE_FACETS are resolved so it should be easier to see the real changes now. 

The changes are all regarding `build_kdtree`, and I don't think I have any further improvements to this area, so it could be reviewed. Thanks.

File: Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:60]: I just convinced myself that there is no need for the "3D equivalent"   as there is a typedef in Static_filters.h

Comment (lrineau) [reply Nef_3/include/CGAL/Nef_3/SNC_k3_tree_traits.h:60]: I was wondering why a `typedef` as sufficient for `Compare_x_3` but a dedicated functor was required for `Compare_x_2`. The difference is:
  - the concept [`CompareX_3`](https://doc.cgal.org/latest/Kernel_23/classKernel_1_1CompareX__3.html) specifies only one overload, that only compares coordinates,
  - whereas the concept [`CompareX_2`](https://doc.cgal.org/latest/Kernel_23/classKernel_1_1CompareX__2.html) has overloads that require comparison of compute values.

**Comment by GilesBathgate** [comment]
@afabri I've increased the scope of the benchmark to include Difference/Intersection/Union. The full results are now generated using the github actions CI. The latest results are here: https://github.com/afabri/Nef_3_benchmark/suites/3621074286/artifacts/87561967

In summary, I get 17% to 18% performance improvement. 

Perhaps the PR is only improving performance for this dataset (spheregrid/shiftedspheregrid), but I would find that very odd.

**Comment by afabri** [comment]
I just edited my comment, where I wrote that master is sometimes faster.   This is only the case for the "(A - B)  union with intersection(A,B)".   This is  a rather artificial testcase where the operations end up to be the input object A,  But  that it is a little bit slower may also come from the fact that these are cascaded operations, that is the output of operations is the input of some more operations.  To combine both scenarios in a single test, was a bad idea.

**Comment by afabri** [comment]
I wrote a just iterative testcase and it shows performance improvements:
```
In master :
$ ./build/Release/test_iterative.exe
47104
a1 : 47104  92160 21.908 sec.
intersection(a1,a2) : 86585  98402 12.817 sec.
a3 - a4 : 103600  120394 14.138 sec.
a5 - a6 : 96910  116952 13.202 sec.
union(a12, a34) : 203892  190133 25.606 sec.
a1234 - a56 : 274639  228382 38.42 sec.

$ ./build/Release/test_iterative.exe
47104
a1 : 47104  92160 17.897 sec.
intersection(a1,a2) : 86585  98402 10.294 sec.
a3 - a4 : 103600  120394 11.082 sec.
a5 - a6 : 96910  116952 10.653 sec.
union(a12, a34) : 203892  190133 20.267 sec.
a1234 - a56 : 274639  228382 31.08 sec.
```
As input I used several copies of the spheregrid, each rotated and shifted.  
image

**Comment by afabri** [comment]
Maybe it is worth to still investigate the case where the branch is slower. What is particular about it is that two inputs share many vertices and faces as they got constructed by a previous Boolean operation.

**Review by afabri — APPROVED** [review]

**Comment by afabri** [comment]
As far I can see it we can remove the label "Tests failing"

**Comment by sloriot** [comment]
Still fully red:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.4-Ic-50/Nef_3/TestReport_Sosno_MSVC-2019-Community-Release.gz

**Comment by GilesBathgate** [comment]
@sloriot Sorry, I assumed the patch by afabri would fix this so didn't test locally. I now fixed in 217ee10 and have the test result:

```
successful configuration
   successful compilation of nary
   successful execution   of nary
   successful compilation of nef_3_from_points
   successful execution   of nef_3_from_points
   successful compilation of Nef_3_problematic_construction
   successful execution   of Nef_3_problematic_construction
   successful compilation of nef_union_error_llvm
   successful execution   of nef_union_error_llvm
   successful compilation of test_nef_3_io_Cartesian
   successful execution   of test_nef_3_io_Cartesian
   successful compilation of test_nef_3_io_Cartesian_Lazy
   successful execution   of test_nef_3_io_Cartesian_Lazy
   successful compilation of test_nef_3_io_EPEC
   successful execution   of test_nef_3_io_EPEC
   successful compilation of test_nef_3_io_EPIC
   successful execution   of test_nef_3_io_EPIC
   successful compilation of test_nef_3_io_Homogeneous
   successful execution   of test_nef_3_io_Homogeneous
   successful compilation of test_nef_3_io_Homogenoeus_Lazy
   successful execution   of test_nef_3_io_Homogenoeus_Lazy
   successful compilation of test_nef_to_soup
   successful execution   of test_nef_to_soup
   successful compilation of test_with_extended_homogeneous
   successful execution   of test_with_extended_homogeneous
   successful compilation of test_with_homogeneous
   successful execution   of test_with_homogeneous
   successful compilation of test_with_lazy
   successful execution   of test_with_lazy
```

**Comment by sloriot** [comment]
Full red line in [Minkowski_sum_3](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.4-Ic-52/Minkowski_sum_3/TestReport_lrineau_Debian-Stable.gz) and [Convex_decomposition](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.4-Ic-52/Convex_decomposition_3/TestReport_gimeno_ArchLinux-clang-Release.gz)

**Comment by GilesBathgate** [comment]
@sloriot Sorry again. It just didn't occur to run the full test suite, locally. Reproduced and fixed in 1918ed3 

P.S the test suite has great coverage, and the assertion immediately pointed me to the problem. Really good.

**Comment by GilesBathgate** [comment]
>And does the fix change anything for the performance tests?

@afabri There is no performance degradation as can be seen in these latest results: https://github.com/afabri/Nef_3_benchmark/suites/3786753639/artifacts/93351812

File: Nef_3/include/CGAL/Nef_3/K3_tree.h
Comment (GilesBathgate) [inline Nef_3/include/CGAL/Nef_3/K3_tree.h:650]: I changed the comparison here from `<2` to `<=2` based on the assumption that 1 vertex per tree cell is overkill. However maybe the code elsewhere assumes there can only be one candidate per cell when for example doing the point location?

**Comment by sloriot** [comment]
Runtime errors for several platforms for `Convex_decomposition_3_Examples` in [CGAL-5.4-Ic-108](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.4-Ic-108.shtml).

File: Nef_3/include/CGAL/Nef_3/K3_tree.h
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:650]: Is this change the culprit for the runtime errors signaled by @sloriot?    I had approved the PR, and it would make sense to not continue working on a PR  by adding further improvements, but to limit it to fixing bugs in the testsuite.

Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/K3_tree.h:650]: @afabri  The change was always in the PR that you approved, I just added the comment about it later in hindsight. That does raise the question of why this is showing up in the testsuite now where it previously passed. Anyway I am not intending to make any further improvements, like you say, limit this to fixing bugs, and make a new PR for anything else. 

When I have some time this weekend I'll investigate, I think I just need to run the testsuite locally, again, but like I say this didn't show up when I ran the testsuite before.

**Comment by GilesBathgate** [comment]
@afabri does 7b8c3ab fix the failing test. If so that's great thanks.

**Comment by afabri** [comment]
It should fix. We were only by minutes to late to put it in the testsuite. Will happen tonight.

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.4-Ic-125

**Pull request merged**
